### ISSUE TITLE
curl error when pusing to apns

### ISSUE THREAD

**Issue by algesiaResidue** [edamov/pushok#175]
Sometimes, We got this error when pusing to apns: Operation timed out after 0 milliseconds with 0 out of 0 bytes received. The frequence is once or twice a day.

**Comment by edamov** [comment]
It is hard to fix such error. Maybe the problem is on apple server side.
The only way I see is to add try/catch block and send request again in case of error. You can implement it on your side.
But also I may try to add this logic in library. Can you somehow store the whole response (status code, headers, body) from APNS and send me?

**Comment by algesiaResidue** [comment]
unfortunatly, curl http code is 0, thus throwing this exception in Client.php line 161, and there is no valid response.

**Comment by algesiaResidue** [comment]
maybe adding a retry function when curl get code 0 is a good idea.

**Comment by dlucian** [comment]
But then, since the whole sending crashes on one timeout, one would need to check which notifications where sent and re-try the sending with just the ones that were not sent, right?

Which can't be done, since `Client::push()` doesn't return the responses anymore, but throws instead. So we don't know which notifications were sent and which not. Or am I missing something?

**Comment by algesiaResidue** [comment]
> But then, since the whole sending crashes on one timeout, one would need to check which notifications where sent and re-try the sending with just the ones that were not sent, right?
> 
> Which can't be done, since `Client::push()` doesn't return the responses anymore, but throws instead. So we don't know which notifications were sent and which not. Or am I missing something?

You can't just throw this exception, but handle it yourself instead. When curl http code got 0, you should add that notification to multi curl handler again to be sent later.

**Comment by dlucian** [comment]
> > But then, since the whole sending crashes on one timeout, one would need to check which notifications where sent and re-try the sending with just the ones that were not sent, right?
> > Which can't be done, since `Client::push()` doesn't return the responses anymore, but throws instead. So we don't know which notifications were sent and which not. Or am I missing something?
> 
> You can't just throw this exception, but handle it yourself instead. When curl http code got 0, you should add that notification to multi curl handler again to be sent later.

No, the `Client` class throws `\Exception`, not me. And when it it's thrown, there's no list of successfully sent notifications, so I don't know which to retry.

**Comment by algesiaResidue** [comment]
You can rewrite this method, and add retry mechanism. I just did this, and the error never affected my pushing project again lol.

**Comment by algesiaResidue** [comment]
Sorry for the misinformation, after adding the retry mechanism ,this issue still bothers me, apns server will continue to deny this notification in next few minutes which means you need retry through this time span.  I've figure out a solution, I recommand you use swoole http2 client and connection pool to maintain these connections. you can refer to https://wiki.swoole.com/zh-cn/#/coroutine_client/http2_client for more information.

**Comment by zhaozhenxiang** [comment]
`
Resolving timed out after 1000 milliseconds
`
`
Failed to connect to api.push.apple.com port 443 after 999 ms
`
`
Connection timed out after 300 milliseconds
`
`
Operation timed out after 235 milliseconds with 0 bytes received
`

I think the problem is with Apple's servers.
I also think using http keepalive will improve the situation.

**Comment by mobielbekeken-thijs** [comment]
In my experience Apple returns a time out when the APNS server is being called too many times in a short period.
When this happens, I let the script sleep for 10 seconds and try again. This works 99/100 times.
The time out can be avoided by not sending too many notifications in a short period.
